databases/Assignment2_answers.sql

```diff
@@ -0,0 +1,316 @@
+
+13. Which manufacturers supply only personal computers?
+
+SELECT distinct manufacturers
+FROM Product
+WHERE Type ='pc'
+
+
+
+with pcs AS (
+    SELECT Manufacturer, Model
+    FROM Product
+    WHERE Type = 'pc'
+),
+
+
+
+
+with not_pcs AS (
+    SELECT Manufacturer
+    FROM Product
+    WHERE Type != 'pc'
+    GROUP BY Manufacturer
+)
+SELECT Product.Manufacturer
+FROM Product
+WHERE Manufacturer NOT IN not_pcs;
+
+ALSO:
+
+select all of the ones that ONLY have one type
+intersect
+SELECT all of the ones that are only type= PC
+
+
+# sub query with Type != pc, outer left join it together ...
+
+EXCEPT; IS NOT IN
+
+HAVING SUM(Type) = 'pc';
+
+HAVING Type = 'pc';
+# HAVING MAX(TYPE) = 'pc';
+# the case where manufacturers sell PCs, but other things as well!
+
+
+# HAVING Type = pc .... i dont think this is right, but why?
+
+
+14. Which manufacturers supply only one type of product?
+SELECT Manufactuer
+FROM Product
+GROUP BY Manufactuer
+HAVING COUNT(DISTINCT Type) = 1;
+
+
+
+15. You want to buy: #think about how to intersect, instead of JOINS? which is more efficient? why?
+• a personal computer with at least 1024 MB of RAM and a CD with 56x, and
+• a laptop with a screen larger than 1500, and
+• a printer cheaper than $300
+all produced by the same manufacturer. Which manufacturers should you consider?
+
+with computers AS (
+    SELECT distinct Manufacturer
+    FROM Computer
+    INNER JOIN Product USING Model
+    WHERE Product.Type = pc AND RAM > 1024 AND CD = 56x;
+),
+laptops AS (
+    SELECT distinct Manufactuer
+    FROM Computer
+    INNER JOIN Product USING Model
+    WHERE Product.Type = laptop AND Computer.Screen > 15;
+),
+printers AS (
+    SELECT distinct Manufacturer
+    FROM Printer
+    INNER JOIN Product USING Model
+    WHERE Price < 300;
+)
+SELECT c.Manufacturer
+FROM computers c
+INNER JOIN laptops l USING Manufacturer
+INNER JOIN printers p USING Manufacturer
+;
+
+
+
+
+# do sequential JOINS happen simultaneously?
+SELECT DISTINCT Product.Manufacturer
+FROM Product
+INNER JOIN Computer USING Model
+WHERE Product.Type = pc AND Computer.RAM > 1024 AND Computer.CD = 56x
+INNER JOIN Computer USING Model
+WHERE Product.Type = laptop AND Computer.Screen > 15
+INNER JOIN Printer
+WHERE Printer.Price < 300
+GROUP BY Product.Manufacturer;
+
+
+
+^ ERROR. not how innerjoins work :(
+
+
+
+
+
+
+
+
+
+16. What HD sizes are common to at least 2 computers?
+SELECT HD
+FROM Computer
+GROUP BY HD
+HAVING COUNT(Model) > 2;
+
+
+17. Which manufacturers sell more than one computer that
+has a processor speed of less than 3 GHz?
+SELECT Product.Manufacturer
+FROM Product
+WHERE Computer.Speed < 3
+INNER JOIN Computer USING Model
+GROUP BY Product.Manufacturer
+HAVING COUNT(Product.Manufacturer) > 1;
+
+
+18. Which manufacturers sell the fastest computers?
+# general case where I do not know fastest speed
+SELECT DISTINCT Product.Manufacturer
+FROM Product
+INNER JOIN Computer USING (Model)
+WHERE Speed = max(Speed)
```

Comment:
This doesn't make sense because you can't aggregate (`MAX`) in the WHERE clause. And you also can't access `Speed` in the HAVING. You'll have to do a subquery for the max speed and JOIN it to each row and then compare.

Or, you can use a window function. Window functions are not supported in all flavours of SQL, but it works well in this case. Look it up if you have time, but otherwise just do the first method